### ISSUE TITLE
Allow assets to be stored in Java module's packages

### ISSUE THREAD

**Issue by sormuras** [AlmasB/FXGL#967]
Today, when bundling FXGL assets in a Java module, one has to use packages that start with `assets.` and open them to FXGL. Like in this [module-info.java](https://github.com/sormuras/toccata/blob/main/com.github.sormuras.toccata/module-info.java):

```java
module com.github.sormuras.toccata {
  requires com.almasb.fxgl.all;
  requires jdk.crypto.ec; // https://github.com/AlmasB/FXGL/issues/910

  exports com.github.sormuras.toccata to com.almasb.fxgl.core;

  opens assets.textures to com.almasb.fxgl.all;
}
```

I'd like to store all assets below the `com.github.sormuras.toccata` "namespace" in order to avoid conflicts with other modules that might bring in their assets. Using the Maven-style directory layout, it could read:

```
src/main/java/
  com/github/sormuras/toccata/ToccataEntityFactory.java
  ...
src/main/resources/
  com/github/sormuras/toccata/assets/textures/stone.png
  ...
```

This would yield a cleaner modular API and avoid split packages:

```java
module com.github.sormuras.toccata {
  requires com.almasb.fxgl.all;
  requires jdk.crypto.ec; // https://github.com/AlmasB/FXGL/issues/910

  exports com.github.sormuras.toccata to com.almasb.fxgl.core;

  opens com.github.sormuras.toccata.assets.textures to com.almasb.fxgl.all;
}
```

Thinking aloud ... a `FXGL.setBasePackageForAssets("com.github.sormuras.toccata")` would be neat, or allowing absolute paths (starting with a `/` character) when loading assets, like `FXGL.texture("/com/github/sormuras/toccata/assets/textures/stone.png")`.

**Comment by AlmasB** [comment]
This proposal looks good. Provided the current behaviour remains unaltered, this would be a good feature to have.

> Thinking aloud ... a FXGL.setBasePackageForAssets("com.github.sormuras.toccata") would be neat

This needs to become a constant value in the context of the engine's lifetime, so it will go to:

```
initSettings(settings) {
    settings.setBasePackageForAssets("com.github.sormuras.toccata")
}
```

All other code, e.g. `texture("stone.png")` should remain unchanged, since the above setting should tell FXGL to look for `assets` in "com.github.sormuras.toccata". This sounds really good -- with just 1 setting, users can easily move around the "assets" directory as needed.

> or allowing absolute paths (starting with a / character) when loading assets, like FXGL.texture("/com/github/sormuras/toccata/assets/textures/stone.png").

This is a separate issue I think, related to #860

**Comment by sormuras** [comment]
Perhaps, the feature needs to be module-aware at runtime. Meaning, a `texture("stone.png)` called in module `foo` expects to load from package `foo.assets.textures` and when called from module `bar` it is package `bar.assets.textures`.

**Comment by AlmasB** [comment]
All of asset loading comes down to [this function](https://github.com/AlmasB/FXGL/blob/dev/fxgl/src/main/kotlin/com/almasb/fxgl/app/services/FXGLAssetLoaderService.kt#L475) to obtain a url. So module-awareness would need to be baked in there somehow?

**Comment by sormuras** [comment]
`Class.getResource(String)` delegates to [ClassLoader.findResource(String,String)](
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ClassLoader.html#findResource(java.lang.String,java.lang.String)) ... which could help here, methinks.

**Comment by AlmasB** [comment]
Ok, I'm happy to look at it in more detail as a separate issue -- currently there's a bunch of assets provided internally by `fxgl.all` module, which we should be able to read. So if we have

```
module foo
stone.png

module bar
stone.png

module fxgl.all
stone.png
```

asset loading first should check the module in which we are running. If it's `foo` we load stone.png from foo. Suppose the file is not there, then we should load stone.png from `fxgl.all`. <-- this behaviour is important.

As for the current issue, would you like to have a go at implementing it? In theory, we only need to modify:

* [Settings](https://github.com/AlmasB/FXGL/blob/dev/fxgl/src/main/kotlin/com/almasb/fxgl/app/Settings.kt) to be able to pass the base package for assets
* And the [getURL](https://github.com/AlmasB/FXGL/blob/dev/fxgl/src/main/kotlin/com/almasb/fxgl/app/services/FXGLAssetLoaderService.kt#L475) function from above.

Otherwise, I'm happy to look at this later in the week.

**Comment by sormuras** [comment]
> As for the current issue, would you like to have a go at implementing it?

Not this week. Already planned my hacking resources to their limit...

**Comment by AlmasB** [comment]
To get the latest build, you can use this:
```
<dependencies>
    <dependency>
        <groupId>com.github.almasb</groupId>
        <artifactId>fxgl</artifactId>
        <version>dev-SNAPSHOT</version>
    </dependency>
</dependencies>

<repositories>
    <repository>
        <id>oss.sonatype.org-snapshot</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
    </repository>
</repositories>
```

Here's [the sample](https://github.com/AlmasB/FXGL/blob/dev/fxgl-samples/src/main/java/sandbox/CustomPathAssetsSample.java) showing how to set a different base package. In that sample, the `testcoin.png` file is located in `src/main/resources/com/almasb/testpackage/assets/textures/testcoin.png`. Please let me know if this works for your use case.

**Comment by sormuras** [comment]
I'll give it a shot in a minute and report back here.

**Comment by sormuras** [comment]
Works great for a single module project.

Double-checking for when two (or more) modules are involved...

**Comment by sormuras** [comment]
...as expected, a single application-wide base package is not working out for when multiple modules need provide their own base package names.

![image](https://user-images.githubusercontent.com/2319838/108509720-e6b5ed80-72bd-11eb-9e6f-2ba1ba24be42.png)

Browse the history of toccata at this commit: https://github.com/sormuras/toccata/tree/97aef7523832e06ce569e06a2f913d3148dd9a2d

TL;DR

The "base" game provides its assets via `com.github.sormuras.toccata.assets` and tells FXGL about it.

Next, it loads its plugins to add their entities to the game.

https://github.com/sormuras/toccata/blob/97aef7523832e06ce569e06a2f913d3148dd9a2d/com.github.sormuras.toccata/com/github/sormuras/toccata/Toccata.java#L66

The following "expansion" code fails to load "ball.png" due to it provides its assets in `com.github.sormuras.toccata.expansion.assets`

https://github.com/sormuras/toccata/blob/97aef7523832e06ce569e06a2f913d3148dd9a2d/com.github.sormuras.toccata.expansion/com/github/sormuras/toccata/expansion/ExpansionToccataPlugin.java#L17

**Comment by sormuras** [comment]
> > or allowing absolute paths (starting with a `/` character) when loading assets, like
> > `FXGL.texture("/com/github/sormuras/toccata/assets/textures/stone.png")`.
> 
> This is a separate issue I think, related to #860

Perhaps, the base package change feature (`settings.setBasePackageForAssets(...)`) should be dropped in favor of the URI loading feature then.
